src/Shared/Microsoft.DotNet.Internal.Health/AzureTableHealthReportProvider.cs

```diff
@@ -0,0 +1,234 @@
+        private static (string instance, string subStatus) ParseRowKey(string rowKey)
+        {
+            var parts = rowKey.Split('|');
+            var subStatus = UnescapeKeyField(parts[1]);
+            if (string.IsNullOrEmpty(parts[0]))
+                return (null, subStatus);
+            return (UnescapeKeyField(parts[0]), subStatus);
+        }
+
+        private static string EscapeKeyField(string value) =>
+            value.Replace(":", "\0")
+                .Replace("|", ":pipe:")
+                .Replace("\\", ":back:")
+                .Replace("/", ":slash:")
+                .Replace("#", ":hash:")
+                .Replace("?", ":question:")
+                .Replace("\0", ":colon:");
+
+        private static string UnescapeKeyField(string value) =>
+            value.Replace(":colon:", "\0")
+                .Replace(":pipe:", "|")
+                .Replace(":back:", "\\")
+                .Replace(":slash:", "/")
+                .Replace(":hash:", "#")
+                .Replace(":question:", "?")
+                .Replace("\0", ":");
+
+        public async Task UpdateStatusAsync(string serviceName, string instance, string subStatusName, HealthStatus status, string message)
+        {
+            string partitionKey = EscapeKeyField(serviceName);
+            string rowKey = GetRowKey(instance, subStatusName);
+
+            async Task Attempt()
+            {
+                HttpContent content = new ByteArrayContent(JsonSerializer.SerializeToUtf8Bytes(
+                    new Entity {Status = status, Message = message},
+                    s_jsonSerializerOptions
+                ));
+                content.Headers.ContentType = MediaTypeHeaderValue.Parse("application/json");
+
+                string requestUri =
+                    $"{_baseUrl}(PartitionKey='{Uri.EscapeDataString(partitionKey)}',RowKey='{Uri.EscapeDataString(rowKey)}'){_sasQuery}";
+                using HttpResponseMessage response = await _client.PutAsync(requestUri, content);
+                response.EnsureSuccessStatusCode();
+            }
+
+            try
+            {
+                await ExponentialRetry.RetryAsync(
+                    Attempt,
+                    e => _logger.LogWarning("Failed to update status for {service}/{subStatus}, retrying",
+                        serviceName,
+                        subStatusName),
+                    e => true
+                );
+            }
+            catch (Exception e)
+            {
+                // Crashing out a service trying to report health isn't useful, log that we failed and move on
+                _logger.LogError(e, "Unable to update health status for {service}/{subStatus}", serviceName, subStatusName);
+            }
+        }
+
+        public async Task<HealthReport> GetStatusAsync(string serviceName, string instance, string subStatusName)
```

Comment:
I'm wondering if this method doesn't make any sense... when would you know the "instance" of something you are trying to get status for? Maybe only the "all status" method makes sense.

Which means I can get rid of the "ForExternalInstance" method on the factory.

@alexperovich thoughts?

Reply:
It probably doesn't make sense to get the status for a specific "instance". That would require already knowing what instances exist, at which point you are already calling the FabricClient, might as well just read the health there.

Reply:
Done. Removed the instance stuff and made the ExponentialRetry stuff injectable/configurable. Otherwise when the tests failed they took like 10 minutes (they just sat in retry loops waiting for the fake server to never answer)